### PR TITLE
golang filter extension: add more methods in the StreamInfo interface.

### LINKED ISSUE #25806: golang filter extension: add more methods in the StreamInfo interface.
These methods could be useful, for the first step.
```
type StreamInfo interface {
	FilterChainName() string
	// Protocol return the request's protocol.
	Protocol() (string, bool)
	// ResponseCode return the response code.
	ResponseCode() (uint32, bool)
	// ResponseCodeDetails return the response code details.
	ResponseCodeDetails() (string, bool)
	// AttemptCount return the number of times the request was attempted upstream.
	AttemptCount() uint32
}
```

Furthermore, we may introduce a common mechanism to get variables from Envoy to Golang, i.e. https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage

### PR DESCRIPTION
**Pull request by doujiang24**
Commit Message: fix #25806
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: contrib/golang/filters/http/source/go/pkg/api/filter.go [hunk 1]
```diff
@@ -58,7 +58,15 @@ type StreamEncoderFilter interface {
...
-	// TODO add more for stream info
+	FilterChainName() string
+	// Protocol return the request's protocol.
+	Protocol() (string, bool)
+	// ResponseCode return the response code.
+	ResponseCode() (uint32, bool)
+	// ResponseCodeDetails return the response code details.
+	ResponseCodeDetails() (string, bool)
+	// AttemptCount return the number of times the request was attempted upstream.
+	AttemptCount() uint32
```
Comment (wangfakang) [inline contrib/golang/filters/http/source/go/pkg/api/filter.go:69]: AttemptCount() (uint32, bool)

same elsewhere

File: contrib/golang/filters/http/source/go/pkg/http/filter.go [hunk 2]
```diff
@@ -99,5 +113,37 @@ type streamInfo struct {
...
+	name, _ := cAPI.HttpGetStringValue(unsafe.Pointer(s.request.req), ValueFilterChainName)
+	return name
+}
+
+func (s *streamInfo) Protocol() (string, bool) {
+	if protocol, ok := cAPI.HttpGetIntegerValue(unsafe.Pointer(s.request.req), ValueProtocol); ok {
+		if name, ok := protocolsIdToName[protocol]; ok {
+			return name, true
+		}
+		panic(fmt.Sprintf("invalid protocol id: %d", protocol))
```
Comment (wangfakang) [inline contrib/golang/filters/http/source/go/pkg/http/filter.go:130]: invalid protocol id is not a fatal error, so it's better to return `"", false` directly.

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 3]
```diff
@@ -753,6 +753,43 @@ CAPIStatus Filter::setTrailer(absl::string_view key, absl::string_view value) {
...
+    *value = state.streamInfo().responseCode().value();
+    break;
+  case EnvoyValue::AttemptCount:
+    if (!state.streamInfo().attemptCount().has_value()) {
+      return CAPIStatus::CAPIValueNotFound;
+    }
+    *value = state.streamInfo().attemptCount().value();
+    break;
+  default:
+    RELEASE_ASSERT(false, absl::StrCat("invalid integer value id: ", id));
```
[the remark this answers is not in the document]
Comment (doujiang24) [reply contrib/golang/filters/http/source/golang_filter.cc:788]: It's impossible to have other ids when people use the Go APIS that we provided. One possible way is that people use the raw C API, that not supported by us.

One major premise is that we assume the Go code could be trusted, as well as we trust the C++ code when implementing an HTTP filter.

So, I prefer to fail earlier, and make the code simpler.

File: contrib/golang/filters/http/source/go/pkg/http/filter.go (hunk 2 above)
Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/http/filter.go:130]: `Http::Protocol` is an enum class, there won't be other values.
https://github.com/envoyproxy/envoy/blob/main/envoy/http/protocol.h#L12

File: contrib/golang/filters/http/source/go/pkg/api/filter.go (hunk 1 above)
Comment (doujiang24) [reply contrib/golang/filters/http/source/go/pkg/api/filter.go:69]: According to the comment: `absl::nullopt if the request was never attempted upstream.`
https://github.com/envoyproxy/envoy/blob/main/envoy/stream_info/stream_info.h#L762

We could simplify the `never` to zero, to make the API usage simpler.

**Review by wangfakang — APPROVED** [review]
Great work. LGTM

**Comment by doujiang24** [comment]
/assign @wbpcode 
Please help to take a look, thanks.

**Comment by wbpcode** [comment]
sorry for the delay. I will take a look this weekend.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**